source/common/upstream/health_checker_impl.cc

```diff
@@ -148,37 +149,63 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
 
 HttpHealthCheckerImpl::HttpStatusChecker::HttpStatusChecker(
     const Protobuf::RepeatedPtrField<envoy::type::v3::Int64Range>& expected_statuses,
+    const Protobuf::RepeatedPtrField<envoy::type::v3::Int64Range>& retriable_statuses,
     uint64_t default_expected_status) {
   for (const auto& status_range : expected_statuses) {
-    const auto start = status_range.start();
-    const auto end = status_range.end();
+    const auto start = static_cast<uint64_t>(status_range.start());
+    const auto end = static_cast<uint64_t>(status_range.end());
 
-    if (start >= end) {
-      throw EnvoyException(fmt::format(
-          "Invalid http status range: expecting start < end, but found start={} and end={}", start,
-          end));
-    }
+    validateRange(start, end, "expected");
 
-    if (start < 100) {
-      throw EnvoyException(fmt::format(
-          "Invalid http status range: expecting start >= 100, but found start={}", start));
-    }
+    expected_ranges_.emplace_back(std::make_pair(start, end));
+  }
 
-    if (end > 600) {
-      throw EnvoyException(
-          fmt::format("Invalid http status range: expecting end <= 600, but found end={}", end));
-    }
+  if (expected_ranges_.empty()) {
+    expected_ranges_.emplace_back(
+        std::make_pair(default_expected_status, default_expected_status + 1));
+  }
+
+  for (const auto& status_range : retriable_statuses) {
+    const auto start = static_cast<uint64_t>(status_range.start());
+    const auto end = static_cast<uint64_t>(status_range.end());
+
+    validateRange(start, end, "retriable");
+
+    retriable_ranges_.emplace_back(std::make_pair(start, end));
+  }
+}
+
+void HttpHealthCheckerImpl::HttpStatusChecker::validateRange(uint64_t start, uint64_t end,
+                                                             absl::string_view range_type) const {
+  if (start >= end) {
+    throw EnvoyException(fmt::format("Invalid http {} status range: expecting start < "
+                                     "end, but found start={} and end={}",
+                                     range_type, start, end));
+  }
 
-    ranges_.emplace_back(std::make_pair(static_cast<uint64_t>(start), static_cast<uint64_t>(end)));
+  if (start < 100) {
+    throw EnvoyException(
+        fmt::format("Invalid http {} status range: expecting start >= 100, but found start={}",
+                    range_type, start));
   }
 
-  if (ranges_.empty()) {
-    ranges_.emplace_back(std::make_pair(default_expected_status, default_expected_status + 1));
+  if (end > 600) {
+    throw EnvoyException(fmt::format(
+        "Invalid http {} status range: expecting end <= 600, but found end={}", range_type, end));
```

Comment:
```suggestion
        "Invalid http {} status range: expecting end < 600, but found end={}", range_type, end));
```

Reply:
This is correct as is. Due to following the half-open semantics, end is not considered in range. So an end of 600 means up to 599 is valid.